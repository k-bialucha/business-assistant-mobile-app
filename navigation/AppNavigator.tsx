import React from 'react';
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs';
import { createStackNavigator } from '@react-navigation/stack';

import DashboardScreen from '../screens/DashboardScreen';
import SalesScreen from '../screens/SalesScreen';
import CostsScreen from '../screens/CostsScreen';
import CostsEntryScreen from '../screens/CostsEntryScreen';
import SalesEntryScreen from '../screens/SalesEntryScreen';

const DashboardStackNavigator = createStackNavigator();

const DashboardNavigator = () => {
  return (
    <DashboardStackNavigator.Navigator>
      <DashboardStackNavigator.Screen
        name="Dashboard"
        component={DashboardScreen}
      />
    </DashboardStackNavigator.Navigator>
  );
};

const SalesStackNavigator = createStackNavigator();

const SalesNavigator = () => {
  return (
    <SalesStackNavigator.Navigator>
      <SalesStackNavigator.Screen name="SalesList" component={SalesScreen} />
      <SalesStackNavigator.Screen
        name="SalesEntry"
        component={SalesEntryScreen}
      />
    </SalesStackNavigator.Navigator>
  );
};

const CostsStackNavigator = createStackNavigator();

const CostsNavigator = () => {
  return (
    <CostsStackNavigator.Navigator>
      <CostsStackNavigator.Screen name="CostsList" component={CostsScreen} />
      <CostsStackNavigator.Screen
        name="CostsEntry"
        component={CostsEntryScreen}
      />
    </CostsStackNavigator.Navigator>
  );
};

const AppTabNavigator = createBottomTabNavigator();

const AppNavigator = () => {
  return (
    <AppTabNavigator.Navigator>
      <AppTabNavigator.Screen name="Dashboard" component={DashboardNavigator} />
      <AppTabNavigator.Screen name="Sales" component={SalesNavigator} />
      <AppTabNavigator.Screen name="Costs" component={CostsNavigator} />
    </AppTabNavigator.Navigator>
  );
};

export default AppNavigator;
