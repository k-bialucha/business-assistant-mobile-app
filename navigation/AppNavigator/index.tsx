import React from 'react';
import { Platform } from 'react-native';

import { Ionicons } from '@expo/vector-icons';
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs';
import { RouteProp } from '@react-navigation/native';
import { StackNavigationProp } from '@react-navigation/stack';

import Colors from '../../theme/Colors';

import CostsNavigator from './CostsNavigator';
import DashboardNavigator from './DashboardNavigator';
import SalesNavigator from './SalesNavigator';
import SettingsNavigator from './SettingsNavigator';

export type RootStackParamList = {
  Dashboard: undefined;
  Sales: undefined;
  Costs: undefined;
  Settings: undefined;
};

export type ScreenProps<RouteName extends keyof RootStackParamList> = {
  navigation: StackNavigationProp<RootStackParamList, RouteName>;
  route: RouteProp<RootStackParamList, RouteName>;
};

const AppTabNavigator = createBottomTabNavigator<RootStackParamList>();

const AppNavigator: React.FC<{}> = () => {
  return (
    <AppTabNavigator.Navigator
      initialRouteName="Dashboard"
      screenOptions={({ route }) => ({
        tabBarIcon: ({ color, size }) => {
          const icons = {
            Dashboard: Platform.OS === 'android' ? 'md-stats' : 'ios-stats',
            Sales:
              Platform.OS === 'android' ? 'md-trending-up' : 'ios-trending-up',
            Costs:
              Platform.OS === 'android'
                ? 'md-trending-down'
                : 'ios-trending-down',
            Settings:
              Platform.OS === 'android' ? 'md-settings' : 'ios-settings',
          };

          return (
            <Ionicons name={icons[route.name]} size={size} color={color} />
          );
        },
      })}
      tabBarOptions={{
        showLabel: false,
        activeTintColor: Colors.navyBlue,
        inactiveTintColor: Colors.silver,
      }}
    >
      <AppTabNavigator.Screen name="Sales" component={SalesNavigator} />
      <AppTabNavigator.Screen name="Dashboard" component={DashboardNavigator} />
      <AppTabNavigator.Screen name="Costs" component={CostsNavigator} />
      <AppTabNavigator.Screen name="Settings" component={SettingsNavigator} />
    </AppTabNavigator.Navigator>
  );
};

export default AppNavigator;
