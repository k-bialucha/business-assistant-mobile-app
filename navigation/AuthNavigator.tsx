import React from 'react';

import { RouteProp } from '@react-navigation/native';
import {
  createStackNavigator,
  StackNavigationProp,
} from '@react-navigation/stack';

import LoginScreen from '../screens/LoginScreen';

export type ParamList = {
  Login: undefined;
};

const StackNavigator = createStackNavigator<ParamList>();

const AuthNavigator = () => {
  return (
    <StackNavigator.Navigator>
      <StackNavigator.Screen name="Login" component={LoginScreen} />
    </StackNavigator.Navigator>
  );
};

export default AuthNavigator;

// export type so screens can get typing
export type NavigationData<RouteName extends keyof ParamList> = {
  navigation: StackNavigationProp<ParamList, RouteName>;
  route: RouteProp<ParamList, RouteName>;
};
