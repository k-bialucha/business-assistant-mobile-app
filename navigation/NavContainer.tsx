import React from 'react';

import { NavigationContainer } from '@react-navigation/native';
import { useSelector } from 'react-redux';

import Loader from '../components/UI/Loader';
import { RequestStatus } from '../redux/auth/types';
import { RootState } from '../redux/rootReducer';
import StartupScreen from '../screens/StartupScreen';

import AppNavigator from './AppNavigator';
import AuthNavigator from './AuthNavigator';

const NavContainer = () => {
  const didTryAutoLogin: boolean = useSelector(
    (state: RootState) => !!state.auth.didTryAutoLogin
  );

  const isAuthenticated: boolean = useSelector(
    (state: RootState) => !!state.auth.token
  );

  const isLoading: boolean = useSelector(
    (state: RootState) => state.auth.requestStatus === RequestStatus.LOADING
  );

  return (
    <>
      <NavigationContainer>
        {isAuthenticated && <AppNavigator />}
        {!isAuthenticated && didTryAutoLogin && <AuthNavigator />}
        {!isAuthenticated && !didTryAutoLogin && <StartupScreen />}
      </NavigationContainer>
      <Loader isLoading={isLoading} />
    </>
  );
};

export default NavContainer;
