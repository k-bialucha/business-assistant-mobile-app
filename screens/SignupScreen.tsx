import React from 'react';
import { ScrollView } from 'react-native';

import { Formik } from 'formik';
import { Button } from 'react-native-elements';
import { useDispatch } from 'react-redux';
import * as Yup from 'yup';

import TextField from '../components/form/TextField';
import { NavigationData } from '../navigation/AuthNavigator';
import { signup } from '../redux/auth/actions';
import Colors from '../theme/Colors';

import { StyledContainer, StyledWideContainer } from './SignupScreen.styled';

const SignupSchema = Yup.object().shape({
  email: Yup.string()
    .email('Invalid email')
    .required('Required'),
  phone: Yup.string()
    .min(9, 'Too Short!')
    .max(12, 'Too Long!'),
  password: Yup.string()
    .min(6, 'Too Short!')
    .required('Required'),
});

type Props = NavigationData<'Signup'>;

const SignupScreen: React.FC<Props> = () => {
  const dispatch = useDispatch();

  return (
    <ScrollView
      keyboardShouldPersistTaps="handled"
      style={{ backgroundColor: Colors.navyBlue }}
    >
      <StyledContainer>
        <Formik
          initialValues={{ email: '', password: '', phone: '' }}
          validationSchema={SignupSchema}
          onSubmit={({ email, password, phone }) => {
            dispatch(signup(email, password, phone));
          }}
        >
          {({
            handleChange,
            handleBlur,
            handleSubmit,
            values,
            errors,
            touched,
          }) => (
            <StyledWideContainer>
              <TextField
                testID="email-input"
                value={values.email}
                placeholder="E-mail"
                onChangeText={handleChange('email')}
                onBlur={handleBlur('email')}
                error={errors.email}
                touched={touched.email}
                keyboardType="email-address"
              />
              <TextField
                testID="phone-input"
                value={values.phone}
                placeholder="Phone (Optional)"
                onChangeText={handleChange('phone')}
                onBlur={handleBlur('phone')}
                error={errors.phone}
                touched={touched.phone}
                keyboardType="phone-pad"
              />
              <TextField
                testID="password-input"
                value={values.password}
                placeholder="Password"
                onChangeText={handleChange('password')}
                onBlur={handleBlur('password')}
                error={errors.password}
                touched={touched.password}
                secureTextEntry
              />
              <Button
                testID="signup-button"
                title="Sign Up"
                buttonStyle={{ backgroundColor: '#ffffff', marginTop: 15 }}
                titleStyle={{ color: Colors.gray }}
                onPress={() => {
                  handleSubmit();
                }}
              />
            </StyledWideContainer>
          )}
        </Formik>
      </StyledContainer>
    </ScrollView>
  );
};

export default SignupScreen;
