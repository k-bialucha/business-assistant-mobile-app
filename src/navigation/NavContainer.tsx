import React, { useEffect } from 'react';
import { Alert } from 'react-native';

import { NavigationContainer } from '@react-navigation/native';
import { useDispatch, useSelector } from 'react-redux';

import Loader from '~/components/UI/Loader';
import { clearErrorState, tryAutoLogin } from '~/redux/auth/actions';
import { RequestStatus } from '~/redux/auth/types';
import { RootState } from '~/redux/rootReducer';

import { AppNavigator } from './AppNavigator';
import AuthNavigator from './AuthNavigator';

const NavContainer = () => {
  const dispatch = useDispatch();

  const { requestStatus, error, isAuthenticated } = useSelector(
    (state: RootState) => state.auth
  );

  const isLoading: boolean = requestStatus === RequestStatus.LOADING;

  useEffect(() => {
    if (error) {
      Alert.alert(error.title, error.message);
      dispatch(clearErrorState());
    }
  }, [error, dispatch]);

  useEffect(() => {
    dispatch(tryAutoLogin());
  }, [dispatch]);

  return (
    <>
      <NavigationContainer>
        {isAuthenticated ? <AppNavigator /> : <AuthNavigator />}
      </NavigationContainer>
      <Loader isLoading={isLoading} />
    </>
  );
};

export default NavContainer;
