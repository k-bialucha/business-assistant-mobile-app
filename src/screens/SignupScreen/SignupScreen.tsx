import React from 'react';
import { ScrollView } from 'react-native';

import { Formik } from 'formik';
import { Button } from 'react-native-elements';
import { useDispatch } from 'react-redux';
import * as Yup from 'yup';

import TextField from '~/components/form/TextField';
import { useAppTranslation } from '~/hooks/useAppTranslation';
import { NavigationData } from '~/navigation/AuthNavigator';
import { signup } from '~/redux/auth/actions';
import Colors from '~/theme/Colors';

import { StyledContainer, StyledWideContainer } from './SignupScreen.styled';

type Props = NavigationData<'Signup'>;

const SignupScreen: React.FC<Props> = () => {
  const dispatch = useDispatch();
  const { t } = useAppTranslation();

  const SignupSchema = Yup.object().shape({
    email: Yup.string()
      .email(t('Invalid Email'))
      .required(t('Required')),
    phone: Yup.string()
      .min(9, t('Too short'))
      .max(12, t('Too long')),
    password: Yup.string()
      .min(6, t('Too short'))
      .required(t('Required')),
  });

  return (
    <ScrollView
      keyboardShouldPersistTaps="handled"
      style={{ backgroundColor: Colors.navyBlue }}
    >
      <StyledContainer>
        <Formik
          initialValues={{ email: '', password: '', phone: '' }}
          validationSchema={SignupSchema}
          onSubmit={({ email, password, phone }) => {
            dispatch(signup(email, password, phone));
          }}
        >
          {({
            handleChange,
            handleBlur,
            handleSubmit,
            values,
            errors,
            touched,
          }) => (
            <StyledWideContainer>
              <TextField
                testID="email-input"
                value={values.email}
                placeholder={t('E-mail')}
                onChangeText={handleChange('email')}
                onBlur={handleBlur('email')}
                error={errors.email}
                touched={touched.email}
                keyboardType="email-address"
              />
              <TextField
                testID="phone-input"
                value={values.phone}
                placeholder={t('Phone (Optional)')}
                onChangeText={handleChange('phone')}
                onBlur={handleBlur('phone')}
                error={errors.phone}
                touched={touched.phone}
                keyboardType="phone-pad"
              />
              <TextField
                testID="password-input"
                value={values.password}
                placeholder={t('Password')}
                onChangeText={handleChange('password')}
                onBlur={handleBlur('password')}
                error={errors.password}
                touched={touched.password}
                secureTextEntry
              />
              <Button
                testID="signup-button"
                title={t('Sign Up')}
                buttonStyle={{ backgroundColor: '#ffffff', marginTop: 15 }}
                titleStyle={{ color: Colors.gray }}
                onPress={() => {
                  handleSubmit();
                }}
              />
            </StyledWideContainer>
          )}
        </Formik>
      </StyledContainer>
    </ScrollView>
  );
};

export default SignupScreen;
