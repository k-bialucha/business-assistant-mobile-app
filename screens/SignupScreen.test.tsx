import React from 'react';

import { shallow } from 'react-native-testing-library';

import { NavigationData } from '../navigation/AuthNavigator';

import SignupScreen from './SignupScreen';

type Props = NavigationData<'Signup'>;

const mockedProps: Props = {
  route: { key: '1234', name: 'Signup' },
  // @ts-ignore
  navigation: {},
};

describe('<SignupScreen />', () => {
  it('matches the snapshot', () => {
    const mockedComp = shallow(<SignupScreen {...mockedProps} />);

    expect(mockedComp).toMatchSnapshot();
  });
});
