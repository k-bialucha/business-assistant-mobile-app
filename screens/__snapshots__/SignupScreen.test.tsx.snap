// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignupScreen /> matches the snapshot 1`] = `
Object {
  "output": <ForwardRef>
    <ForwardRef>
      <TextField
        onChangeText={[Function]}
        placeholder="E-mail"
        value=""
      />
      <TextField
        onChangeText={[Function]}
        placeholder="Phone (Optional)"
        value=""
      />
      <TextField
        onChangeText={[Function]}
        placeholder="Password"
        secureTextEntry={true}
        value=""
      />
      <ForwardRef(Themed.Button)
        buttonStyle={
          Object {
            "backgroundColor": "#ffffff",
            "marginTop": 15,
          }
        }
        onPress={[Function]}
        title="Sign Up"
        titleStyle={
          Object {
            "color": "#5E6D8C",
          }
        }
      />
    </ForwardRef>
  </ForwardRef>,
}
`;
